src/runtime_src/core/pcie/driver/linux/xocl/subdev/hwmon_sdm.c

```diff
@@ -0,0 +1,517 @@
+                                sizeof(char) * strlen(sysfs_name), GFP_KERNEL);
+	strcpy(iter->dev_attr.attr.name, sysfs_name);
+	iter->dev_attr.attr.mode = S_IRUGO;
+	iter->dev_attr.show = hwmon_sensor_show;
+	iter->index = repo_id | (field_id << 4) | (buf_index << 8) | (len << 20);
+
+	sysfs_attr_init(&iter->dev_attr.attr);
+	err = device_create_file(sdm->hwmon_dev, &iter->dev_attr);
+	if (err) {
+		iter->dev_attr.attr.name = NULL;
+		xocl_err(&sdm->pdev->dev, "unabled to create sensors_list0 hwmon sysfs file ret: 0x%x",
+				 err);
+	}
+
+	return err;
+}
+
+static uint8_t sdr_get_id(uint8_t repo_type)
+{
+	int id = 0;
+
+	switch(repo_type) {
+		case SDR_TYPE_BDINFO:
+			id = 0;
+			break;
+		case SDR_TYPE_TEMP:
+			id = 1;
+			break;
+		case SDR_TYPE_VOLTAGE:
+			id = 2;
+			break;
+		case SDR_TYPE_CURRENT:
+			id = 3;
+			break;
+		case SDR_TYPE_POWER:
+			id = 4;
+			break;
+		case SDR_TYPE_QSFP:
+			id = 5;
+			break;
+		case SDR_TYPE_VPD_PCIE:
+			id = 6;
+			break;
+		case SDR_TYPE_IPMIFRU:
+			id = 7;
+			break;
+		case SDR_TYPE_CSDR_LOGDATA:
+			id = 8;
+			break;
+		case SDR_TYPE_VMC_LOGDATA:
+			id = 9;
+			break;
+		default:
+			id = -1;
+			break;
+	}
+
+	return id;
+}
+
+static void parse_sdr_info(char *in_buf, struct xocl_hwmon_sdm *sdm,
+                           bool create_sysfs)
+{
+	bool create = false;
+	uint8_t status;
+	int buf_index, err;
+	uint8_t remaining_records, completion_code, repo_type, repo_id;
+	uint8_t name_length, name_type_length;
+	uint8_t val_len, value_type_length, threshold_support_byte;
+	uint8_t bu_len, sensor_id, base_unit_type_length, unit_modifier_byte;
+	uint32_t name_index, ins_index, max_index = 0, avg_index = 0;
+
+	completion_code = in_buf[SDR_COMPLETE_IDX];
+
+	xocl_dbg(&sdm->pdev->dev, "Parsing SDR Repository: received completion_code: 0x%x",
+			 completion_code);
+
+	if(completion_code != SDR_CODE_OP_SUCCESS)
+	{
+		if(completion_code == SDR_CODE_NOT_AVAILABLE)
+			xocl_err(&sdm->pdev->dev, "Error: SDR Code Not Available");
+		else if(completion_code == SDR_CODE_OP_FAILED)
+			xocl_err(&sdm->pdev->dev, "Error: SDR Code Operation Failed");
+		else if(completion_code == SDR_CODE_FLOW_CONTROL_READ_STALE)
+			xocl_err(&sdm->pdev->dev, "Error: SDR Code Flow Control Read Stale");
+		else if(completion_code == SDR_CODE_FLOW_CONTROL_WRITE_ERROR)
+			xocl_err(&sdm->pdev->dev, "Error: SDR Code Flow Control Write Error");
+		else if(completion_code == SDR_CODE_INVALID_SENSOR_ID)
+			xocl_err(&sdm->pdev->dev, "Error: SDR Code Invalid Sensor ID");
+		else
+			xocl_err(&sdm->pdev->dev, "Failed in sending SDR Repository command");
+		return;
+	}
+
+	repo_type = in_buf[SDR_REPO_IDX];
+	repo_id = sdr_get_id(repo_type);
+	if (repo_id < 0) {
+		xocl_err(&sdm->pdev->dev, "SDR Responce has INVALID REPO TYPE: %d", repo_type);
+		return;
+	}
+
+	buf_index = SDR_NUM_BYTES_IDX + 1;
+
+	remaining_records = in_buf[SDR_NUM_REC_IDX];
+
+	while(remaining_records > 0)
+	{
+		sensor_id = in_buf[buf_index++];
```

Comment:
actually, we treat in_buf as a u8 array, so the buf_index should never beyond the size.
So we should check buf_index*sizeof(u8) should < in_buf[SDR_NUM_BYTES_IDX]

Reply:
in_buf[SDR_NUM_BYTES_IDX] will have max value of 0xFF (255 bytes). So, not sure if it can be used for index checking. Since each sensor record may have more bytes than in_buf[SDR_NUM_BYTES_IDX].